### PR TITLE
[TT-13669] Add pre-commit, pre-push hooks

### PR DESCRIPTION
**Pull request by jeffy-mathew**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-13669" title="TT-13669" target="_blank">TT-13669</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Add pre commit hooks to run task lint</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Task
        Task
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Dev</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td>-</td>
    </tr>
  </table>
</details>

---

## Description
This PR adds pre commit hook scripts to the project.
A task command `setup` is also added, so that a new contributor can have the dependencies installed and pre commit hook setup locally.

## Related Issue
https://tyktech.atlassian.net/browse/TT-13669

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Enhancement, Documentation

___

### **Description**
- Introduced a pre-commit hook script (`scripts/pre-commit.sh`) to enforce linting before commits.
- Added a `setup_hooks` task in `.taskfiles/deps/Taskfile.yml` for setting up pre-commit hooks.
- Updated `Taskfile.yml` with a new `setup` task to streamline project initialization, including dependencies and hooks.
- Enhanced `CONTRIBUTING.md` with instructions on using Task commands for setup and testing, emphasizing the new pre-commit hook setup.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>pre-commit.sh</strong><dd><code>Add pre-commit script for linting enforcement</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

scripts/pre-commit.sh

<li>Added a pre-commit script to run linting tasks before commits.<br> <li> Ensures commits are aborted if linting fails.<br> <li> Provides feedback on linting success or failure.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6733/files#diff-daf6c6cbb9aa58db0628121a2e695ff3ec13927c9b3641aebc8f2be89d231552">+14/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Add task for pre-commit hook setup</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.taskfiles/deps/Taskfile.yml

<li>Added a <code>setup_hooks</code> task to set up pre-commit hooks.<br> <li> Updated the <code>default</code> task to include <code>setup_hooks</code>.<br> <li> Defined commands for copying and enabling the pre-commit script.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6733/files#diff-bd2d25e8a19db2d54ea34295acdc50b5402b7469eaf3e16b083f3e66dbe4fd9d">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>Taskfile.yml</strong><dd><code>Add setup task for project initialization</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

Taskfile.yml

<li>Added a <code>setup</code> task to install dependencies and pre-commit hooks.<br> <li> Linked <code>setup</code> task to <code>deps:setup_hooks</code> for hook setup.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6733/files#diff-cd2d359855d0301ce190f1ec3b4c572ea690c83747f6df61c9340720e3d2425e">+6/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Documentation</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>CONTRIBUTING.md</strong><dd><code>Update contributing guide with Task commands</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

CONTRIBUTING.md

<li>Updated documentation to include Task commands for setup and testing.<br> <li> Added instructions for setting up pre-commit hooks using <code>task setup</code>.<br> <li> Improved clarity on using Task for local development.<br>

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/6733/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055">+10/-2</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by buger** [comment]
<p> A JIRA Issue ID is missing from your branch name, PR title and PR description! 🦄</p>
<p>Your branch: chores/pre-commit-lint</p>
<p>Your PR title: Add pre-commit hooks</p>
<p>Your PR description: 

## Description
This PR adds pre commit hook scripts to the project.
A task command `setup` is also added, so that a new contributor can have the dependencies installed and pre commit hook setup locally.

## Related Issue






## Motivation and Context



## How This Has Been Tested






## Screenshots (if appropriate)

## Types of changes



- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist





- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why
</p>
<p>If this is your first time contributing to this repository - welcome!</p>
<hr />
<p>Please refer to <a href="https://github.com/cleartax/jira-lint">jira-lint</a> to get started.
<p>Without the JIRA Issue ID in your branch name you would lose out on automatic updates to JIRA via SCM; some GitHub status checks might fail.</p>
Valid sample branch names:

  ‣ feature/shiny-new-feature--mojo-10'
  ‣ 'chore/changelogUpdate_mojo-123'
  ‣ 'bugfix/fix-some-strange-bug_GAL-2345'

File: scripts/pre-commit.sh [hunk 1]
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh
+
```
Comment (titpetric) [inline scripts/pre-commit.sh:2]: `exec task lint` should be the shortest option here? maybe it needs bash for `exec`;

**Comment by buger** [comment]
<p>I'm a bot and I 👍 this PR title. 🤖</p>

File: .taskfiles/deps/Taskfile.yml [hunk 2]
```diff
@@ -47,6 +48,14 @@ tasks:
     cmds:
       - go install github.com/fatih/faillint@latest
 
+  lefthook:
+    internal: true
+    desc: 'Install lefthook'
+    status:
+      - type lefthoo
```
Comment (titpetric) [inline .taskfiles/deps/Taskfile.yml:55]:
```suggestion
      - type lefthook
```

File: lefthook.yml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+pre-commit:
```
Comment (titpetric) [inline lefthook.yml:1]: Pre-push missing, trailing \n missing in yamls (overall)

File: scripts/pre-commit.sh (hunk 1 above)
Comment (titpetric) [reply scripts/pre-commit.sh:2]: unused with lefthook

File: Taskfile.yml [hunk 4]
```diff
@@ -95,3 +96,37 @@ tasks:
...
     cmds:
       - go test -count 1 -v -run=TestXTykGateway_Lint ./apidef/oas/
+
+  setup:
+    desc: "Setup the project including dependencies and git hooks using lefthooks"
+    cmds:
+      - task: deps:default
+      - lefthook install
+
+  simple-build:
```
Comment (titpetric) [inline Taskfile.yml:106]: you could move all or of these new targets from here on under .taskfiles/lint.yml right?

File: lefthook.yml [hunk 5]
```diff
@@ -0,0 +1,8 @@
...
+  commands:
+    task:
+      run: task pre-commit
```
Comment (titpetric) [inline lefthook.yml:4]:
```suggestion
      run: task lint:pre-commit
```

**Pull request merged**